doc/api/net.md

```diff
@@ -629,6 +632,21 @@ For [IPC][] connections, available `options` are:
   See [Identifying paths for IPC connections][]. If provided, the TCP-specific
   options above are ignored.
 
+For both types, available `options` include:
+
+* `onread` {Object} - If specified, incoming data is stored in a single `buffer`
+  and passed to the supplied `callback` when data arrives on the socket.
+  Note: this will cause the streaming functionality to not provide any data,
+  however events like `'error'`, `'end'`, and `'close'` will still be emitted
+  as normal and methods like `pause()` and `resume()` will also behave as
+  expected.
+  * `buffer` {Buffer|Uint8Array} - A reusable chunk of memory to use for storing
+    incoming data.
+  * `callback` {Function} This function is called for every chunk of incoming
+    data. Two arguments are passed to it: the number of bytes written to
+    `buffer` and a reference to `buffer`. Return `false` from this function to
+    implicitly `pause()` the socket.
+
```

Comment:
Do I read correctly that it's up to the user to size `buffer` accordingly to the expected incoming data and that on every callback the buffer is partially rewritten? If that's the case, I'd suggest making the docs a bit more clear on that part.

Reply:
That is correct. Do you have any suggestions as to what should be added?

Reply:
It depends. What is the typical deciding factor on when a new chunk is written? packet MTU?

Reply:
Whenever a packet is received? The reading mechanism is exactly the same as before, the only difference here is that instead of node allocating a new Buffer for every chunk/packet, you can reuse the same Buffer, avoiding the overhead of Buffer allocation and additional GCs.

Reply:
Sorry for being unclear, I'm not really familar on the chunking mechanism or the fact that it would be packet-based. So, would you recommend a buffer size >= MTU? Assuming a user might also want to support jumbo frames would 9000 would be a good recommendation?

Reply:
You can use whatever size you want, it just depends on your needs. The current stream-based implementation simply tries to allocate whatever libuv suggests, which is [currently 64KB](https://github.com/nodejs/node/blob/8772da6ea890afbe9088b71edf98fd3023090b00/deps/uv/src/unix/stream.c#L1138). So you could use that same size if you wanted, but having a custom size is really a secondary benefit. The main goal with this PR however is to avoid continuous allocations.

Reply:
Ok. Maybe add a sentence like:

> The size of this buffer needs to be set to a value that is equal or greater to the maximum expected chunk size, which for a typical request is currently 65535.

This is of course assuming that if a bigger chunk arrives, the implementation will misbehave.

Reply:
It doesn't *need* to be any particular size. It's just a tradeoff between memory usage and number of `read()` syscalls (and JS<->C++ boundary crossing). The larger the Buffer, the fewer `read()` syscalls are required to read `n` bytes from the socket.

Reply:
I see. Doesn't that make it a implementation detail that does not need to be exposed to user code then. Why not remove the buffer argument and set a reasonable default size and just emit a reference to it in the callback?

Reply:
It's not an implementation detail, it's explicit on purpose. Being able to explicitly set a Buffer allows for greater flexibility. Not only does it allow you to choose whatever size you want, but you can also use a chunk of memory you may already have allocated.